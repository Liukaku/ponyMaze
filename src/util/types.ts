export interface MazeProps {
  key: number;
  walls: Cardinals[];
  isPony: boolean;
  isDomokun: boolean;
  isExit: boolean;
  height: number;
  width: number;
}

export type Cardinals = "north" | "south" | "east" | "west";

export interface GoalLocations {
  pony: number | null;
  domokun: number | null;
  endPoint: number | null;
}
