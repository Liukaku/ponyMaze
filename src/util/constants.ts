export type XY = "height" | "width" | "difficulty";

export type MLPCharacters =
  | "Twilight Sparkle"
  | "Rainbow Dash"
  | "Pinkie Pie"
  | "Rarity"
  | "Applejack"
  | "Fluttershy";

export const MAX = 25;
export const MIN = 15;

export const MlpChars: Record<string, MLPCharacters> = {
  lowResTwiSp: "Twilight Sparkle",
  lowResRainDa: "Rainbow Dash",
  lowResPinPi: "Pinkie Pie",
  lowResRari: "Rarity",
  lowResAppJa: "Applejack",
  lowFluShy: "Fluttershy",
};

export const btnStyle =
  "shadow-sm shadow-black select-none max-h-12 w-auto cursor-pointer text-black items-center flex px-2 my-3 border-2 bg-gray-300 hover:bg-gray-400 active:bg-gray-600 border-b-gray-500 border-l-gray-500 border-r-gray-100 border-t-gray-100 active:border-t-gray-700 active:border-r-gray-700 active:border-l-gray-200 active:border-b-gray-200";
