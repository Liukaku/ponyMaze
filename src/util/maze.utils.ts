import { MazeProps, Cardinals, ContructMazeProps } from "./types";

export function constructMaze(
  mazeDataParam: ContructMazeProps
): MazeProps[][] | undefined {
  if (Object.keys(mazeDataParam).length === 0) return;
  const rows: MazeProps[][] = [];

  let id = 0;
  const { height, width, walls, pony, domokun, endPoint } = mazeDataParam;

  for (let i = 0; i < height; i++) {
    const row: MazeProps[] = [];
    for (let j = 0; j < width; j++) {
      const blockData: MazeProps = {
        key: id,
        walls: walls[id],
        isPony: id === pony[0] ? true : false,
        isDomokun: id === domokun[0] ? true : false,
        isExit: id === endPoint[0] ? true : false,
        height: height,
        width: width,
      };
      if (i === height - 1) {
        blockData.walls.push("south");
      }
      if (j === width - 1) {
        blockData.walls.push("east");
      }

      row.push(blockData);
      id++;
    }
    rows.push(row);
  }
  return rows;
}
