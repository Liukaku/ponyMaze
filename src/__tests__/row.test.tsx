import "@testing-library/jest-dom";
import React from "react";
import renderer from "react-test-renderer";
import { MazeProps } from "../util/types";
import Row from "../components/row";

describe("Row tests", () => {
  it("renders a row with the given props", () => {
    const walls = rowMock({ howMany: 15 });
    const aRow = renderer
      .create(<Row walls={walls} pony={"lowResTwiSp"} />)
      .toJSON();
    expect(aRow).toMatchSnapshot();
  });
});

function rowMock(overrides: { howMany: number }): MazeProps[] {
  const walls: MazeProps[] = [];
  for (let i = 0; i < overrides.howMany; i++) {
    walls.push({
      key: i,
      walls: [],
      width: 15,
      height: 15,
      isPony: i === 0,
      isDomokun: i === 1,
      isExit: i === 2,
    });
  }

  return walls;
}
