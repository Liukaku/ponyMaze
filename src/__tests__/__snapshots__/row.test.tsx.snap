// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row tests renders a row with the given props 1`] = `
<div
  className="flex justify-center"
>
  <div
    className="text-white bg-pink-200/80"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  >
    <img
      alt=""
      className="mx-auto"
      data-nimg="1"
      decoding="async"
      height={50}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fponies%2FlowResTwiSp.png&w=128&q=75"
      srcSet="/_next/image?url=%2Fponies%2FlowResTwiSp.png&w=64&q=75 1x, /_next/image?url=%2Fponies%2FlowResTwiSp.png&w=128&q=75 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={50}
    />
  </div>
  <div
    className="text-white bg-red-800/80"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  >
    <img
      alt=""
      className="mx-auto"
      data-nimg="1"
      decoding="async"
      height={100}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fdomo.png&w=256&q=100"
      srcSet="/_next/image?url=%2Fdomo.png&w=128&q=100 1x, /_next/image?url=%2Fdomo.png&w=256&q=100 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={100}
    />
  </div>
  <div
    className="text-white bg-yellow-200/80"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  >
    <img
      alt=""
      className="mx-auto"
      data-nimg="1"
      decoding="async"
      height={100}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fdoor.png&w=256&q=100"
      srcSet="/_next/image?url=%2Fdoor.png&w=128&q=100 1x, /_next/image?url=%2Fdoor.png&w=256&q=100 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={100}
    />
  </div>
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
  <div
    className="text-white"
    style={
      {
        "height": "6vmin",
        "width": "6vmin",
      }
    }
  />
</div>
`;
